### PR TITLE
*: bump go to 1.20 and k8s deps to 0.27.2

### LINKED ISSUE #581: Breaks v1.27.x clusters
**What happened?**:
Upgraded kubernetes to v1.27.x. Various k8s-clients started crashing (helm, k9s, metallb, ...). Seems like prom-adapter returns malformed metrics (see https://github.com/kubernetes/kubernetes/pull/116603). Uninstalling prom-adapter fixes the crashes.

**Environment**:
- prometheus-adapter version: v0.10.0
- prometheus version: v2.43.1
- Kubernetes version (use `kubectl version`): v1.27.1
- Cloud provider or hardware configuration: kubeadm

**Steps to reproduce**
- Install kuberntest cluster v1.27.x
- Install kube-prom
- Install prometheus-adapter (from helm chart, with --set prometheus.url=http://kube-prom-kube-prometheus-prometheus)
- try accessing cluster with k9s version <= v0.27.3
- or try installing a helm chart
- or restart metallb-speaker

### PR DESCRIPTION
**Pull request by dgrisonnet**
This PR bumps the various dependency and add support for OpenAPI v3.

Fixes #581

### CODE REVIEW

**Comment by dgrisonnet** [comment]
This requires https://github.com/kubernetes/test-infra/pull/29683.

**Comment by booleanbetrayal** [comment]
@dgrisonnet - Built an image off of this branch to test K8s v1.27 compatibility. Seems to work, except I see this regression again: https://github.com/kubernetes-sigs/prometheus-adapter/issues/525

Had to hot-patch the Helm produced manifest from the project to remove the `'--logtostderr=true'` line it generates. 

Possibly broken in a prior commit?

File: go.mod
Comment (booleanbetrayal) [inline go.mod:3]: this should probably be updated in the Makefile as well

Comment (dgrisonnet) [reply go.mod:3]: good catch :+1:

**Comment by booleanbetrayal** [comment]
> This flag alongside others were removed from Kubernetes components in 1.26 and since prometheus-adapter is reusing code from the kube-apiserver, the same applies here. [kubernetes/enhancements#2845](https://github.com/kubernetes/enhancements/issues/2845)
> 
> I'll update the manifests in this repo to reflect that, but the helm chart will have to be fixed separately.

Thanks for that explanation!

**Review by olivierlemasle — APPROVED** [review]
/lgtm

**Comment by justenwalker** [comment]
FWIW I confirmed this change worked for me. I did my own build of the docker images and re-packaged the helm chart (removing the `--logtostderr=true` flag)

**Comment by dgrisonnet** [comment]
#589 is required to support k8s 1.27

**Comment by dgrisonnet** [comment]
I added the OpenAPI v3 support to this PR directly since it requires client updates, but they require OpenAPI v3 support for the tests to pass.

**Comment by booleanbetrayal** [comment]
> Not sure why it worked for me then. I'm running `v0.6.3` (Helm Chart `v3.10.0`)
> 
> I updated to the latest commit today and it continues to work.

Have also been running this branch's changes for the past couple of weeks without crash.

**Comment by dgrisonnet** [comment]
Now it should be all good, but before my change, if you didn't touch you the OpenAPI v3 feature gate on a 1.27 cluster, it wasn't working properly.

**Comment by justenwalker** [comment]
@dgrisonnet

>  before my change, if you didn't touch you the OpenAPI v3 feature gate on a 1.27 cluster, it wasn't working properly.

I'm deploying on EKS so I do not have access to these feature gates (so they are likely the defaults) - maybe that's why it worked?

**Comment by justenwalker** [comment]
@dgrisonnet is there any additional feedback or testing I can provide to validate this PR? Or is this waiting on approval from someone else?

**Comment by colinrgodsey** [comment]
also stuck on an EKS 1.27 upgrade here. any new developments here on what's needed to move forward with this PR ?

**Comment by dgrisonnet** [comment]
I am just waiting for a review from @olivierlemasle if possible

**Comment by Aym3nTN** [comment]
Can @booleanbetrayal review it?

**Comment by dgrisonnet** [comment]
I'd prefer having someone from kubernetes-sig signing it off, but if it still hasn't made any move next week, I'll merge it and create a new release.

**Comment by justenwalker** [comment]
@dgrisonnet seems unlikely we'll get that review from @olivierlemasle  ; IMO since we've got some independent tests from myself and @booleanbetrayal and its relatively straight forward my vote would be to merge.

Only one note is that we're currently on Go 1.20.6 so it might be good to make that small tweak just to include the security fixes if its an easy enough change. https://github.com/kubernetes-sigs/prometheus-adapter/pull/586/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R11

**Comment by justenwalker** [comment]
@dgrisonnet thanks for closing this out! How does the helm chart for this get built; is that automatic?

The current app version available is v0.10.0

```
$ helm search hub prometheus-adapter
URL                                                     CHART VERSION   APP VERSION     DESCRIPTION
https://artifacthub.io/packages/helm/prometheus...      4.2.0           v0.10.0         A Helm chart for k8s prometheus adapter
...
```

**Comment by dgrisonnet** [comment]
I don't know, the chart is maintained separately.

**Comment by colinrgodsey** [comment]
@justenwalker helm chart PR here: https://github.com/prometheus-community/helm-charts/pull/3641

**Comment by justenwalker** [comment]
> @justenwalker helm chart PR here: [prometheus-community/helm-charts#3641](https://github.com/prometheus-community/helm-charts/pull/3641)

Nice! Thanks @colinrgodsey

**Pull request merged**
